components/proxy/src/main/java/com/hotels/styx/routing/handlers/PathPrefixRouter.java

```diff
@@ -47,42 +43,61 @@
-import static java.util.stream.Collectors.toList;
+import static java.util.Collections.singletonList;
+import static java.util.Objects.requireNonNull;
 
 /**
  * Makes a routing decision based on a request path prefix.
- *
+ * <p>
  * Chooses a destination according to longest matching path prefix.
  * The destination can be a routing object reference or an inline definition.
  */
-public class PathPrefixRouter {
-    public static final Schema.FieldType SCHEMA = object(
-            field("routes", list(object(
-                    field("prefix", string()),
-                    field("destination", routingObject()))
-            ))
-    );
-
-    private final ConcurrentSkipListMap<String, RoutingObject> routes = new ConcurrentSkipListMap<>(
-                    comparingInt(String::length)
-                            .reversed()
-                            .thenComparing(naturalOrder())
-    );
+public class PathPrefixRouter implements RoutingObject {
+    private final Iterable<PrefixRoute> routes;
 
-    PathPrefixRouter(List<Pair<String, RoutingObject>> routes) {
-        routes.forEach(entry -> this.routes.put(entry.key(), entry.value()));
+    PathPrefixRouter(Iterable<PrefixRoute> routes) {
+        this.routes = routes;
     }
 
-    public Optional<RoutingObject> route(LiveHttpRequest request) {
+    @Override
+    public Eventual<LiveHttpResponse> handle(LiveHttpRequest request, HttpInterceptor.Context context) {
         String path = request.path();
 
-        return routes.entrySet().stream()
-                .filter(entry -> path.startsWith(entry.getKey()))
-                .findFirst()
-                .map(Map.Entry::getValue);
+        for (PrefixRoute route : routes) {
+            if (path.startsWith(route.prefix)) {
+                return route.routingObject.handle(request, context);
+            }
+        }
+
+        return Eventual.error(new NoServiceConfiguredException(path));
+    }
+
+    @Override
+    public CompletableFuture<Void> stop() {
+        List<CompletableFuture<Void>> stopFutures = new ArrayList<>();
+        routes.forEach(prefixRoute -> stopFutures.add(prefixRoute.routingObject.stop()));
+        return CompletableFuture.allOf(stopFutures.toArray(new CompletableFuture[stopFutures.size()]));
+    }
+
+    private static class PrefixRoute implements Comparable<PrefixRoute> {
+        private final String prefix;
+        private final RoutingObject routingObject;
+
+        PrefixRoute(String prefix, RoutingObject routingObject) {
+            this.prefix = requireNonNull(prefix);
+            this.routingObject = requireNonNull(routingObject);
+        }
+
+        @Override
+        public int compareTo(@NotNull PrefixRoute o) {
+            if (prefix.length() > o.prefix.length()) {
+                return -1;
+            } else if (prefix.length() < o.prefix.length()) {
+                return 1;
+            } else {
+                return prefix.compareTo(o.prefix);
+            }
+        }
```

Comment:
Please look if you can use the `PathTrie` class (https://github.com/HotelsDotCom/styx/blob/master/components/server/src/main/java/com/hotels/styx/server/PathTrie.java) instead of this ordered list.

The ordered list was copy-pasted from the old monolithic application router code. It is beyond me why it was originally implemented like that. Could have been something to do with origin reloading. But this is just a guess.

I originally used the ordered list approach here to ensure identical behaviour with the old application router. I didn't have time to deal with potential incompatibilities.

Using the `PathTrie`:
* Might be more performant for larger configurations because it is based on hash lookups instead of linear search.
* Eliminates this ordered list as a yet another alternative path prefix lookup implementation.
* IIRC is bit more flexible in terms of it mapping capabilities.

Reply:
`PathTrie` is an interesting class but not a suitable drop in here since it will match each path section as a whole whilst this class is designed to match each character. In this routing object you can match `/sa1` and `sa2` with the prefix `/sa` whilst you cannot with `PathTrie`.

Reply:
I suspected it might not be 100% compatible.

However I do believe this class is not matching each character by design. Instead it is an unintended consequence of the chosen implementation.

Reply:
The [Styx manual](https://github.com/HotelsDotCom/styx/blob/master/docs/user-guide/configure-origins.md) for the old application router merely says that:

> path: the beginning of a (relative) URI used by Styx to determine which backend service to route a request to.

So what do you make of it? Would you think of "beginning of an URI" being a proper parent URI path, or is it just a string arbitrarily chopped off at some point?

Reply:
The attempt here is to do a simple cleanup and refactor of the code without any functional change. I hope to make it easier to read and easier to refactor later. The behavior should stay exactly the same.